lib/src/revset.rs

```diff
@@ -1566,6 +1566,42 @@ fn sort_intersection_by_key<St: ExpressionState, T: Ord>(
     sort_intersection_helper(base, expression, get_key(expression), get_key)
 }
 
+/// Push `ancestors(x)` and `~ancestors(x)` down (to the left) in intersections.
+/// All `~ancestors(x)` will be moved before `ancestors(x)`, since negated
+/// ancestors can be converted to ranges. All other negations are moved to the
+/// right, since these negations can usually be evaluated better as differences.
+fn sort_negations_and_ancestors<St: ExpressionState>(
+    expression: &Rc<RevsetExpression<St>>,
+) -> TransformedExpression<St> {
+    transform_expression_bottom_up(expression, |expression| match expression.as_ref() {
+        RevsetExpression::Intersection(expression1, expression2) => {
+            #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+            enum AncestorsOrder {
```

Comment:
nit: let's move type definition to the (outer) function scope, so the match arms look concise.